cmd/auth-portal/main.go

```diff
@@ -1,64 +1,154 @@
+)
+
+const CookieSigningSecretLength = 24
+
+func main() {
+	cryptoSeed()
 
 	flag.Set("logtostderr", "true")
 
-	configFile := os.Getenv("CONFIG")
-	if configFile != "" {
-		fmt.Fprintf(os.Stderr, "Reading config file %q", configFile)
-
-		yamlBytes, err := ioutil.ReadFile(configFile)
-		if err != nil {
-			fmt.Fprintf(os.Stderr, "error reading config file %q: %v\n", configFile, err)
-			os.Exit(1)
-		}
-
-		jsonBytes, err := yaml.ToJSON(yamlBytes)
-		if err != nil {
-			fmt.Fprintf(os.Stderr, "error converting YAML config file %q: %v\n", configFile, err)
-			os.Exit(1)
-		}
-
-		if err := json.Unmarshal(jsonBytes, &o); err != nil {
-			fmt.Fprintf(os.Stderr, "error parsing YAML config file %q: %v\n", configFile, err)
-			os.Exit(1)
-		}
-	}
-	flag.StringVar(&o.Listen, "listen", o.Listen, "host/port on which to listen")
+	// TODO(componentconfig-q): Some parameters we don't really want configurable, because
+	// we expect to be running in a container.  But maybe they would be useful for people
+	// that want to run the code differently, so they probably warrant a flag or an env var.
+	// Thoughts?
+	listen := ":8080"
+	flag.StringVar(&listen, "listen", listen, "host/port on which to listen")
+	staticDir := "/webapp"
+	flag.StringVar(&staticDir, "static-dir", staticDir, "location of static directory")
 
 	flag.Parse()
 
-	err := run(&o)
+	err := run(listen, staticDir)
 	if err != nil {
 		fmt.Fprintf(os.Stderr, "unexpected error: %v\n", err)
 		os.Exit(1)
 	}
+
+	os.Exit(0)
 }
 
-func run(o *portal.Options) error {
-	p, err := portal.NewHTTPServer(o)
+func run(listen string, staticDir string) error {
+	k8sClient, err := buildK8sClient()
+	if err != nil {
+		return err
+	}
+
+	name := "auth"
+
+	namespaceBytes, err := ioutil.ReadFile("/var/run/secrets/kubernetes.io/serviceaccount/namespace")
+	if err != nil {
+		return fmt.Errorf("error reading namespace from %q: %v", "/var/run/secrets/kubernetes.io/serviceaccount/namespace", err)
+	}
+	namespace := string(namespaceBytes)
+
+	apiContext, err := api.NewAPIContext(os.Getenv("API_VERSIONS"))
+	if err != nil {
+		return fmt.Errorf("error initializing API: %v", err)
+	}
+
+	componentconfiginstall.Install(apiContext.GroupFactoryRegistry, apiContext.Registry, apiContext.Scheme)
+
+	configDecoder := apiContext.Codecs.UniversalDecoder()
+
+	configReader := &configreader.ManagedConfiguration{
+		Decoder: configDecoder,
+	}
+
+	//configFile := os.Getenv("CONFIG")
+	//if configFile != "" {
+	//	err := configReader.Read(configFile)
+	//	if err != nil {
+	//		return fmt.Errorf("error reading config file %q: %v\n", configFile, err)
+	//	}
+	//}
+
+	configObj, err := configReader.ReadFromKubernetes(k8sClient, namespace, name)
+	if err != nil {
+		return fmt.Errorf("error reading configuration: %v", err)
+	}
+
+	// TODO(componentconfig-q): Should we deal with v1alpha1 or unversioned when we own the API?
+	// (I guess the same question with our User objects)
```

Comment:
I think of conversion this way:

The apiserver needs to canonicalize the resource for storage purposes, and convert to/from that representation for the client. Today, we do that conversion by using the internal (unversioned) representation as the hub, but we've talked about eliminating the internal rep for a long time (https://github.com/kubernetes/kubernetes/issues/8190#issuecomment-114139999).

Originally, both defaulting and validation were originally unversioned. Now defaulting is versioned but validation is not. We need to change validation to be versioned (https://github.com/kubernetes/kubernetes/issues/8550). At that point, most of the apiserver's logic would be on the versioned reps, though admission control is TBD.

All other components of the system are clients of the API. (Let's ignore static pods for now.) Ideally, they'd just use the versioned reps, like regular clients (https://github.com/kubernetes/kubernetes/issues/20193). That changes how they'd transition to a new API version, but I think that's a good thing.

So what should components do with versioned configuration? IMO, they should convert it to the canonical versioned representation.